File: android-agent/src/test/java/io/opentelemetry/android/AndroidResourceTest.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android;

import static io.opentelemetry.android.internal.tools.RumConstants.RUM_SDK_VERSION;
import static io.opentelemetry.semconv.ResourceAttributes.DEVICE_MANUFACTURER;
import static io.opentelemetry.semconv.ResourceAttributes.DEVICE_MODEL_IDENTIFIER;
import static io.opentelemetry.semconv.ResourceAttributes.DEVICE_MODEL_NAME;
import static io.opentelemetry.semconv.ResourceAttributes.OS_DESCRIPTION;
import static io.opentelemetry.semconv.ResourceAttributes.OS_NAME;
import static io.opentelemetry.semconv.ResourceAttributes.OS_TYPE;
import static io.opentelemetry.semconv.ResourceAttributes.OS_VERSION;
import static io.opentelemetry.semconv.ResourceAttributes.SERVICE_NAME;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.when;

import android.app.Application;
import android.content.pm.ApplicationInfo;
import android.os.Build;
import io.opentelemetry.sdk.resources.Resource;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Answers;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class AndroidResourceTest {

    String appName = "robotron";
    String rumSdkVersion = BuildConfig.OTEL_ANDROID_VERSION;
    String osDescription =
            new StringBuilder()
                    .append("Android Version ")
                    .append(Build.VERSION.RELEASE)
                    .append(" (Build ")
                    .append(Build.ID)
                    .append(" API level ")
                    .append(Build.VERSION.SDK_INT)
                    .append(")")
                    .toString();

    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
    Application app;

    @Test
    void testFullResource() {
        ApplicationInfo appInfo = new ApplicationInfo();
        appInfo.labelRes = 12345;

        when(app.getApplicationContext().getApplicationInfo()).thenReturn(appInfo);
        when(app.getApplicationContext().getString(appInfo.labelRes)).thenReturn(appName);

        Resource expected =
                Resource.getDefault()
                        .merge(
                                Resource.builder()
                                        .put(SERVICE_NAME, appName)
                                        .put(RUM_SDK_VERSION, rumSdkVersion)
                                        .put(DEVICE_MODEL_NAME, Build.MODEL)
                                        .put(DEVICE_MODEL_IDENTIFIER, Build.MODEL)
                                        .put(DEVICE_MANUFACTURER, Build.MANUFACTURER)
                                        .put(OS_NAME, "Android")
                                        .put(OS_TYPE, "linux")
                                        .put(OS_VERSION, Build.VERSION.RELEASE)
                                        .put(OS_DESCRIPTION, osDescription)
                                        .build());

        Resource result = AndroidResource.createDefault(app);
        assertEquals(expected, result);
    }

    @Test
    void testProblematicContext() {
        when(app.getApplicationContext().getApplicationInfo())
                .thenThrow(new SecurityException("cannot do that"));
        when(app.getApplicationContext().getResources()).thenThrow(new SecurityException("boom"));

        Resource expected =
                Resource.getDefault()
                        .merge(
                                Resource.builder()
                                        .put(SERVICE_NAME, "unknown_service:android")
                                        .put(RUM_SDK_VERSION, rumSdkVersion)
                                        .put(DEVICE_MODEL_NAME, Build.MODEL)
                                        .put(DEVICE_MODEL_IDENTIFIER, Build.MODEL)
                                        .put(DEVICE_MANUFACTURER, Build.MANUFACTURER)
                                        .put(OS_NAME, "Android")
                                        .put(OS_TYPE, "linux")
                                        .put(OS_VERSION, Build.VERSION.RELEASE)
                                        .put(OS_DESCRIPTION, osDescription)
                                        .build());

        Resource result = AndroidResource.createDefault(app);
        assertEquals(expected, result);
    }
}

File: core/src/test/java/io/opentelemetry/android/AndroidResourceTest.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android

import android.content.Context
import android.content.SharedPreferences
import android.content.pm.ApplicationInfo
import android.os.Build
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.mockk
import io.mockk.slot
import io.opentelemetry.api.common.AttributeKey
import io.opentelemetry.api.common.AttributeKey.stringKey
import io.opentelemetry.sdk.resources.Resource
import io.opentelemetry.sdk.resources.ResourceBuilder
import io.opentelemetry.kotlin.semconv.TelemetryAttributes
import io.opentelemetry.kotlin.semconv.AndroidAttributes
import io.opentelemetry.kotlin.semconv.AppAttributes.APP_INSTALLATION_ID
import io.opentelemetry.kotlin.semconv.DeviceAttributes
import io.opentelemetry.kotlin.semconv.IncubatingApi
import io.opentelemetry.kotlin.semconv.OsAttributes
import io.opentelemetry.kotlin.semconv.ServiceAttributes.SERVICE_NAME
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import java.util.UUID

internal class AndroidResourceTest {
    private val appName: String = "robotron"
    private val prefsName: String = "opentelemetry-android"
    private val rumSdkVersion: String = BuildConfig.OTEL_ANDROID_VERSION
    private val installId: String = "install-id"
    private val osDescription: String =
        "Android Version " +
            Build.VERSION.RELEASE +
            " (Build " +
            Build.ID +
            " API level " +
            Build.VERSION.SDK_INT +
            ")"

    @RelaxedMockK
    private lateinit var ctx: Context
    private lateinit var expectedResourceBuilder: ResourceBuilder
    private lateinit var appInfo: ApplicationInfo

    @OptIn(IncubatingApi::class)
    @BeforeEach
    fun setUp() {
        MockKAnnotations.init(this)
        every { ctx.getSharedPreferences(prefsName, 0) } returns
            mockk {
                every {
                    getString(
                        APP_INSTALLATION_ID,
                        null,
                    )
                } returns installId
            }

        appInfo =
            ApplicationInfo().apply {
                labelRes = 12345
            }

        every { ctx.applicationContext.applicationInfo } returns appInfo
        every { ctx.applicationContext.getString(appInfo.labelRes) } returns appName

        expectedResourceBuilder =
            Resource
                .builder()
                .put(SERVICE_NAME, appName)
                .put(TelemetryAttributes.TELEMETRY_SDK_VERSION, rumSdkVersion)
                .put(DeviceAttributes.DEVICE_MODEL_NAME, Build.MODEL)
                .put(DeviceAttributes.DEVICE_MODEL_IDENTIFIER, Build.MODEL)
                .put(DeviceAttributes.DEVICE_MANUFACTURER, Build.MANUFACTURER)
                .put(OsAttributes.OS_NAME, "Android")
                .put(OsAttributes.OS_TYPE, "linux")
                .put(OsAttributes.OS_VERSION, Build.VERSION.RELEASE)
                .put(
                    AndroidAttributes.ANDROID_OS_API_LEVEL,
                    Build.VERSION.SDK_INT.toString(),
                ).put(OsAttributes.OS_DESCRIPTION, osDescription)
                .put(APP_INSTALLATION_ID, installId)
    }

    @Test
    fun testFullResource() {
        assertResourceMatches()
    }

    @Test
    fun `fall back to nonLocalizedLabel if needed`() {
        appInfo =
            ApplicationInfo().apply {
                labelRes = 0
                nonLocalizedLabel = "shim sham"
            }
        every { ctx.applicationContext.applicationInfo } returns appInfo

        assertResourceMatches(
            extraAttributes = mapOf(stringKey(SERVICE_NAME) to "shim sham"),
        )
    }

    @Test
    fun testProblematicContext() {
        every { ctx.applicationContext.applicationInfo } throws SecurityException("cannot do that")
        every { ctx.applicationContext.resources } throws SecurityException("boom")

        assertResourceMatches(
            extraAttributes = mapOf(stringKey(SERVICE_NAME) to "unknown_service:android"),
        )
    }

    @OptIn(IncubatingApi::class)
    @Test
    fun `test install id generated if none available`() {
        val slot = slot<String>()
        val editor = mockk<SharedPreferences.Editor>(relaxed = true)

        every { ctx.getSharedPreferences(prefsName, 0) } returns
            mockk {
                every {
                    getString(
                        APP_INSTALLATION_ID,
                        null,
                    )
                } returns null
                every { edit() } returns editor
            }

        every {
            editor.putString(
                APP_INSTALLATION_ID,
                capture(slot),
            )
        } returns editor

        assertResourceMatches(
            resource = AndroidResource.createDefault(ctx),
            extraAttributes = mapOf(stringKey(APP_INSTALLATION_ID) to slot.captured),
        )
        assertNotNull(UUID.fromString(slot.captured))
    }

    private fun assertResourceMatches(
        resource: Resource = AndroidResource.createDefault(ctx),
        extraAttributes: Map<AttributeKey<*>, String> = emptyMap(),
    ) {
        extraAttributes.forEach { entry ->
            expectedResourceBuilder.put(entry.key.key, entry.value)
        }
        val expected = Resource.getDefault().merge(expectedResourceBuilder.build())
        assertEquals(expected, resource)
    }
}
